docs/guides.md

```diff
@@ -78,27 +78,58 @@ Before we jump into writing some exchanges, there are a couple of
 patterns and limitations that always remain the same when writing
 an exchange.
 
+### Forward Operation Streams. Return OperationResult Streams.
+
 For reference, this is a basic template for an exchange:
 
 ```js
 const noopExchange = ({ client, forward }) => {
-  return operations$ => {
-    return forward(operations$);
+  return operation$ => { // <-- The ExchangeIO function
+    const operationResult$ = forward(operations$);
+    return operationResult$;
+  };
+};
+```
+
+By convention, variables ending with `$` are streams. Each exchange's `ExchangeIO` function receives a stream containing an [`Operation`](https://formidable.com/open-source/urql/docs/api/#operation-type) object (`operation$`) and must return a stream of [`OperationResult`](https://formidable.com/open-source/urql/docs/api/#operationresult-type) object(s)(`operationResult$`).
+
+#### Forward and Return Composition
+
+When you create a client and pass it an array of exchanges, Urql composes them left-to-right into a single exchange using [`composeExchanges`](https://github.com/FormidableLabs/urql/blob/master/src/exchanges/compose.ts). `composeExchanges` orchestrates stream forward and return between exchanges. Here's the noopExchange in context:
+
+```js
+import { Client, dedupeExchange, fetchExchange } from 'urql';
+
+const noopExchange = ({ client, forward }) => {
+  // ExchangeIO function receives Operation object stream from dedupeExchange
+  return operation$ => {
+    // noopExchange forwards Operation object stream to fetchExchange,
+    // and receives a stream with an OperationResult object created by fetchExchange
+    const operationResult$ = forward(operations$);
+    // then returns the OperationResult stream to dedupeExchange
+    return operationResult$;
   };
 };
+
+const client = new Client({
+  exchanges: [dedupeExchange, noopExchange, fetchExchange],
+});
 ```
 
-In this form the exchange doesn't do anything yet.
-When you create a client and pass it an array of exchanges, all exchanges
-will be composed together into a single one. They will each be called with
-an object that contains the `client` itself, and a `forward` function which
-is the next `ExchangeIO` function.
+#### Forward + Return = Request + Response
```

Comment:
I think this section is overall more confusing than helpful. It takes about handlers and ordering, but that’s potentially confusing and factually a little off since we’re dealing with streams.

The example isn’t doing anything so this example isn’t too helpful.
Let’s maybe remove this section since it doesn’t add more explanation, even compared with the deleted section 🤔

Reply:
I agree that this section is "factually off".  It especially wouldn't make sense without fetchExchange on the end making a network request.  I also haven't thought of a better way to address the underlying reason for adding this section yet;  Forward's mismatching semantics and behavior were confusing at the beginning.  The word forward doesn't imply returning a different shape, or returning at all. Calling `forward` is (loosely) like passing a chicken  to `promise.then`, and receving a [turducken](https://en.wikipedia.org/wiki/Turducken) in the next call. That unexpected behavior was confusing when learning.  `getResults`, or `pullResults` semantics would more closely match forward's actual behavior, making it easier to learn.

That said, I didn't want to suggest a name change to a broadly used core function while still learning, hence adding this section.

Reply:
Maybe the example above, showing forward in context, with comments, is sufficient for now, so I'll delete this section.  

If you ever want to deprecate the name `forward`, you'll have my full support. 😁

Reply:
When explaining this, what’s missing here is the hint that `forward` is just the next exchange’s `ExchangeIO` function hence the name.

You pass it a stream of operations you want to _forward_ and you receive the next exchange’s results stream in return.

Maybe a simple explanation like this can go a long way at explaining this.

We also have diagrams from a workshop that I could modify to be included here 🤔

- https://advanced-graphql-workshop.netlify.com/55 (This is very different from urql but can easily be modified to convey a similar amount of information)
- https://advanced-graphql-workshop.netlify.com/65
- https://advanced-graphql-workshop.netlify.com/64

Reply:
I think saying "`forward` calls the next exchange's `ExchangeIO` function" is accurate enough early in the exchange learning process.

"Accurate enough" because my brain choked briefly on the sentence "forward is just the next exchange’s ExchangeIO function".  There are many lower level concepts to parse when reading it as a learner, like an exchange being a function, and each argument in that function, and argument type, and which ones are relevant to the sentence, and operation order, and stream order, and return values.  That's just unpacking the sentence itself.  Saying `forward` _is_ `ExchangeIO` adds the cognitive load of dual names for the same concept.  That's unnecessarily confusing when already remembering the other stuff.  Saying "`forward` calls the next exchange's `ExchangeIO` function" is cognitively easier to keep straight, hence accurate enough.

I can add "`forward` calls the next exchange's `ExchangeIO` function" to the contextualized `noopExchange` example if you're cool with it.

Or... you could always rename the argument `nextExchangeIO`. 😁

Like the video!

Reply:
Contextual noopExchange example expanded.

Some gaps in my own knowledge stood out in the process of explaining pipeline's ExchangIO function role, so I added each step in the process to hopefully prevent others' gaps in understanding.